src/combinatorics/simple_graph/triangle/basic.lean

```diff
@@ -0,0 +1,150 @@
+
+section triangle_free
+
+/-- `G.triangle_free` means that `G` has no triangles. -/
+def triangle_free : Prop := ∀ t, ¬ G.is_n_clique 3 t
+
+variables {G H}
+
+lemma triangle_free_bot : (⊥ : simple_graph α).triangle_free :=
+begin
+  rintro t ⟨ht₁, ht₂⟩,
+  have : 1 < t.card,
+  { rw ht₁,
+    norm_num },
+  rw finset.one_lt_card at this,
+  obtain ⟨a, ha, b, hb, hab⟩ := this,
+  exact ht₂ ha hb hab,
+end
+
+lemma triangle_free.le (h : G ≤ H) : H.triangle_free → G.triangle_free :=
+forall_imp $ λ s, mt $ is_n_clique.le h
+
+end triangle_free
+
+section triangle_finset
+variables (G) [fintype α] [decidable_eq α] [decidable_rel G.adj] {a b c : α}
+
+/-- The triangles in a graph as a finset. -/
+def triangle_finset : finset (finset α) := (powerset_len 3 univ).filter $ G.is_n_clique 3
+
+lemma mem_triangle_finset_iff (s : finset α) : s ∈ G.triangle_finset ↔ G.is_n_clique 3 s :=
+by rw [triangle_finset, mem_filter, mem_powerset_len, and_iff_right (subset_univ _),
+  and_iff_right_of_imp is_n_clique.card_eq]
+
+lemma triple_mem_triangle_finset_iff :
+  {a, b, c} ∈ G.triangle_finset ↔ G.adj a b ∧ G.adj a c ∧ G.adj b c :=
+begin
+  rw [mem_triangle_finset_iff, is_n_clique_iff],
+  simp only [coe_insert, coe_singleton, set.pairwise_insert_of_symmetric G.symm,
+    set.pairwise_singleton, true_and, set.mem_insert_iff, set.mem_singleton_iff,
+    forall_eq_or_imp, forall_eq, ne.def],
+  split,
+  { rintro ⟨h, hbc, hab, hac⟩,
+    refine ⟨hab _, hac _, hbc _⟩;
+    { rintro rfl,
+      simp only [insert_idem, insert_singleton_comm, insert_singleton_self_eq] at h,
+      exact h.not_lt (nat.lt_succ_iff.2 $ card_insert_le _ _) } },
+  { rintro ⟨hab, hac, hbc⟩,
+    refine ⟨_, λ _, hbc, λ _, hab, λ _, hac⟩,
+    rw card_eq_three,
+    exact ⟨_, _, _, G.ne_of_adj hab, G.ne_of_adj hac, G.ne_of_adj hbc, rfl⟩ }
+end
+
+lemma mem_triangle_finset_iff' :
+  s ∈ G.triangle_finset ↔ ∃ a b c, G.adj a b ∧ G.adj a c ∧ G.adj b c ∧ s = {a, b, c} :=
+begin
+  refine ⟨λ h, _, _⟩,
+  { obtain ⟨a, b, c, -, -, -, rfl⟩ := card_eq_three.1 ((G.mem_triangle_finset_iff s).1 h).1,
+    refine ⟨a, b, c, _⟩,
+    rw triple_mem_triangle_finset_iff at h,
+    tauto },
+  { rintro ⟨a, b, c, hab, hbc, hca, rfl⟩,
+    exact G.triple_mem_triangle_finset_iff.2 ⟨hab, hbc, hca⟩ }
+end
+
+@[simp] lemma triangle_finset_eq_empty_iff : G.triangle_finset = ∅ ↔ G.triangle_free :=
+by simp_rw [triangle_free, eq_empty_iff_forall_not_mem, mem_triangle_finset_iff]
+
+alias triangle_finset_eq_empty_iff ↔ _ simple_graph.triangle_free.triangle_finset
+
+attribute [protected] triangle_free.triangle_finset
+
+variables {G} [decidable_rel H.adj]
+
+lemma triangle_finset_mono (h : G ≤ H) : G.triangle_finset ⊆ H.triangle_finset :=
+monotone_filter_right _ $ λ _, is_n_clique.le h
+
+end triangle_finset
+
+section triangle_free_far
+variables [fintype α] {G H} {ε δ : 𝕜}
+
+open_locale classical
+
+/-- A simple graph is `ε`-triangle-free far if one must remove a density of `ε` edges to make it
```

Comment:
```suggestion
/-- A simple graph is `ε`-triangle-free far if one must remove a proportion of `ε` edges to make it
```
how about this?

Reply:
This is misleading (although maybe not more than mine). It's not a proportion of edges but a proportion of *possible* edges.

Reply:
How about we just be more specific here then? 
"A simple graph `G` is `ε`-far from triangle free if one must remove at least `ε (card α)^2` edges to make it triangle free."